filesystem/flysystem/FlysystemUrlPlugin.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+namespace SilverStripe\Filesystem\Flysystem;
+
+use League\Flysystem\AwsS3v2\AwsS3Adapter;
+use League\Flysystem\Filesystem;
+use League\Flysystem\FilesystemInterface;
+use League\Flysystem\PluginInterface;
+use Oneup\FlysystemBundle\Adapter\LocalWithHost;
+
+
+/**
+ * Allows urls for files to be exposed
+ *
+ * Credit to https://github.com/SmartestEdu/FlysystemPublicUrlPlugin
+ */
+class FlysystemUrlPlugin implements PluginInterface {
+	
+    /**
+     * @var Filesystem adapter
+     */
+    protected $adapter;
+
+    public function setFilesystem(FilesystemInterface $filesystem) {
+        $this->adapter = $filesystem->getAdapter();
+    }
+
+    public function getMethod() {
+        return 'getPublicUrl';
+    }
+
+    /**
+	 * Generate public url
+	 * 
+     * @param string $path
+     * @return string The full url to the file
+     */
+    public function handle($path) {
+		// Default adaptor
+		if($this->adapter instanceof AssetAdaptor) {
+			return $this->adapter->getPublicUrl($path);
+		}
+
+		// Check S3 adaptor
+        if (class_exists('League\Flysystem\AwsS3v2\AwsS3Adapter')
```

Comment:
This is simply imported from a third party library which I extended for our own purposes.
https://github.com/SmartestEdu/FlysystemPublicUrlPlugin

Unfortunately, flysystem limits extensions (plugins) to be added only to the filesystem object, so there's no way of actually adding the getPublicURL Extension itself to the backend adapter.

I'm really waiting for the flysystem devs to give us an official API for getting public urls; At the moment it's just not possible without lots of hard coding (as in this case).